app/db/interface.py

```diff
@@ -156,33 +158,50 @@ def _query_payments(self, payment_attributes):
-                if key in new_data:
-                   raise DataValidationError('Invalid payment: body of request contained bad or no data')
-            #if existing_user_id != data['user_id']:
-            #    raise DataValidationError('Invalid payment: Changes to user_id field not allowed')
-            type = new_data['payment_type']
-            detail = new_data['details']
-            if bool(re.search(r'\d', detail['user_name'])) == False:
-                valid = True
-            if type == 'credit' or type == 'debit':
-                card_number = detail['card_number']
-                expires_date = detail['expires']
-                if bool(re.match('^[0-9]+$', card_number)) and (len(card_number) == 16):
-                    datetime.strptime(expires_date, '%m/%Y')
-                    valid_detail = True
-            elif new_data['is_linked'] != None and type == 'paypal':
-                valid_detail = True
-        except KeyError as e:
-            raise DataValidationError('Invalid payment: missing ' + e.args[0])
-        except TypeError as e:
-            raise DataValidationError('Invalid payment: body of request contained bad or no data')
-        return valid & valid_detail
+
+        if user_id and int(new_payment['user_id']) != user_id:
+            raise DataValidationError('Error: You cannot modify the field: user_id')
+
+        for key in self.NONUPDATABLE_PAYMENT_FIELDS :
+            if key in new_payment:
+               raise DataValidationError('Error: You cannot modify the field: %s' % key)
+
+        payment_type = new_payment['payment_type']
+        user_name = new_payment['details']['user_name']
+
+        if bool(re.match("^[A-Za-z\-' ]{0,50}$", user_name)):
+            
+            if payment_type == 'credit' or payment_type == 'debit':
+                card_number = new_payment['details']['card_number']
+                card_type = new_payment['details']['card_type']
+                expires = new_payment['details']['expires']
+                
+                if bool(re.match('^[0-9]{16}$', card_number)):
+                    
+                    if bool(re.match("^[A-Za-z ]{0,10}$", card_type)):
+                        
+                        if bool(re.match('^[0-9]{2}/[0-9]{4}$', expires)):
+                            month = int(expires[:2])
+                            year = int(expires[3:])
+                            
+                            if year > datetime.now().year:
+                                valid = True
+                            elif year == datetime.now().year and month >= datetime.now().month:
+                                valid = True
+
+            elif payment_type == 'paypal':
+                user_email = new_payment['details']['user_email']
+                
+                if bool(re.match('^[\W\w]+@[\W\w]+\.[A-Za-z]{2,3}$', user_email)):
+                    valid = True
+
+        return valid
```

Comment:
Just as a general API note - it's cool for us to issue a blanket `valid` `not valid` to make this work, but it would be better for the consumer of the API to be given a specific message telling them what exactly was invalid.  Definitely not strictly necessary here, but just some food for thought :)